### PR TITLE
 SONARKT-154 Rule S5846: Empty lines should not be tested with regex MULTILINE flag

### PR DESCRIPTION
**Pull request by margarita-nedzelska-sonarsource**

### CODE REVIEW

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/EmptyLineRegexCheck.kt [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+            }
+            is KtDotQualifiedExpression -> {
+                getStringInRegexFind(parent.selectorExpression!!, bindingContext)
+                    ?.let { if (it.canBeEmpty(bindingContext)) listOf(it) else emptyList()} ?: emptyList()
+            }
+            else -> {
+                emptyList()
+            }
+        }) to (callExpression.parent as? KtExpression)?.getResolvedCall(bindingContext)?.getReceiverExpression()
+    }
```
Comment (johann-beleites-sonarsource) [inline sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/EmptyLineRegexCheck.kt:158]: The `when` blocks in these three methods seem to be pretty much the same. I think we can simplify this significantly.

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/EmptyLineRegexCheck.kt [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+    
+    private fun extractArgument(resolvedCall: ResolvedCall<out CallableDescriptor>): KtExpression? {
+        val arguments = resolvedCall.valueArgumentsByIndex
+        return if (arguments == null
+            || arguments.isEmpty()
+            || arguments[0] == null
+            || arguments[0] !is ExpressionValueArgument
+        ) null
+        else (arguments[0] as ExpressionValueArgument).valueArgument?.getArgumentExpression()
```
Comment (johann-beleites-sonarsource) [inline sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/EmptyLineRegexCheck.kt:186]: Can you add some curly braces here? With the line breaks in the condition, it looks a bit unorthodox, and using curly braces might help, I believe.

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/EmptyLineRegexCheck.kt [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+            is KtNameReferenceExpression -> {
+                val runtimeStringValue = deparenthesized.predictRuntimeStringValue(bindingContext)
+                runtimeStringValue?.isEmpty()
+                    ?: deparenthesized.findAllUsages {
+                        (it.parent as? KtExpression)?.getResolvedCall(bindingContext) matches STRING_IS_EMPTY
+                    }.isEmpty()
+            }
+            else ->
+                predictRuntimeStringValue(bindingContext)?.isEmpty() ?: false
+        }
```
Comment (johann-beleites-sonarsource) [inline sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/EmptyLineRegexCheck.kt:200]: It took me a little to understand what this method does and what it looks for - WDYT of a (KDoc) comment to give a very brief explanation of what this method does (and potentially how)?

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/EmptyLineRegexCheck.kt [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+                    .mapNotNull { it.getParentResolvedCall(bindingContext) }
+                    .filter { it matches STRING_REPLACE && it.getReceiverExpression()?.canBeEmpty(bindingContext) == true }
+                    .firstOrNull()
+                    ?.let { kotlinFileContext.reportIssue(it.getFirstArgumentExpression()!!, MESSAGE, listOf(kotlinFileContext.secondaryOf(callExpression))) }
+            }
+            else -> {
+                val resolvedCall = callExpression.getParentResolvedCall(bindingContext)
+                if (resolvedCall matches STRING_REPLACE
+                    && resolvedCall?.getReceiverExpression()?.canBeEmpty(bindingContext) == true) {
+                    kotlinFileContext.reportIssue(resolvedCall.getFirstArgumentExpression()!!, MESSAGE)
```
Comment (johann-beleites-sonarsource) [inline sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/EmptyLineRegexCheck.kt:217]: This looks a bit confusing on first sight - can you either put the condition on one line or the `) {` on a separate line?

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/EmptyLineRegexCheck.kt [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+    var visitedStart = false
+    var visitedEndAfterStart = false
+    var containEmptyLine = false
+    override fun visitSequence(tree: SequenceTree) {
+        val items = tree.items.stream()
+            .filter { item: RegexTree ->
+                !isNonCapturingWithoutChild(item)
+            }
+            .collect(Collectors.toList())
+        if (items.size == 1 && items[0].`is`(RegexTree.Kind.CAPTURING_GROUP)) {
```
Comment (johann-beleites-sonarsource) [inline sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/EmptyLineRegexCheck.kt:236]: No strong opinion; if you want, you can use `.first()` to replace `[0]`

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/api/ApiExtensions.kt [hunk 6]
```diff
@@ -390,6 +390,43 @@ fun KtNameReferenceExpression.findPreviousUsages(
...
+        searchStartNode.getParentOfType<KtBlockExpression>(false)
+            ?.collectDescendantsOfType<KtNameReferenceExpression> { it.getReferencedName() == name }
+            ?.let { usages ->
+                for (usage in usages) {
+                    if (predicate(usage)) {
+                        acc.add(usage)
+                    }
+                }
+            }
+    }
```
Comment (johann-beleites-sonarsource) [inline sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/api/ApiExtensions.kt:428]: These and the `findPreviousUsages` fun look super similar - can we generalize them?

File: sonar-kotlin-plugin/src/main/resources/org/sonar/l10n/kotlin/rules/kotlin/S5846.html [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+<pre>
+static final Pattern p = Pattern.compile("^$", Pattern.MULTILINE);
+
+boolean containsEmptyLines(String str) {
+    return p.matcher(str).find() || str.isEmpty();
+}
+
+// ...
+System.out.println(containsEmptyLines("a\n\nb")); // correctly prints 'true'
+System.out.println(containsEmptyLines("")); // also correctly prints 'true'
```
[the remark this answers is not in the document]
Comment (margarita-nedzelska-sonarsource) [reply sonar-kotlin-plugin/src/main/resources/org/sonar/l10n/kotlin/rules/kotlin/S5846.html:42]: Yep, added examples. We will anyway updated it before the release

**Review by johann-beleites-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**